### DISCUSSION TITLE
HIGH / LOW 演算子

### DISCUSSION THREAD

**Discussion by SuperTurboZ** [AILight/AILZ80ASM#143] [Q&A] [answered]
一部のアセンブラで使える HIGH / LOW 演算子のサポートが無いようですが、何か理由があるのでしょうか？
支障がなく簡単に追加できるようであれば、サポートしていただけると他アセンブラとの互換性が向上するのではないかと思いました。

ところで、@H@Lサンプルなどのreadmeのいくつかのリンクが切れてしまっているようです。
[@H@Lサンプル](https://github.com/AILight/AILZ80ASM/blob/main/AILZ80ASM.Test/Test/TestLB_EQU_Test/Test.Z80)

```
                                ;*** AILZ80ASM *** Z-80 Assembler, version 0.9.9.19, LST:Full:4
000000 0000                         org 0
                                ;   
       1234                     label1  equ 0x1234
                                ;
                                ;現状仕様、互換性維持が必要なケースでは避けたい。
000000 0000 2612             7      ld  h,label1.@HIGH
000002 0002 2E34             7      ld  l,label1.@LOW
                                ;普段はこう書くことが多い（~ と同様の単項演算子）
000004 0004 **** E0004 ****  7      ld  h,high label1
000006 0006 **** E0004 ****  7      ld  l,low label1
                                ;互換性は保てるが可読性が下がる
000008 0008 2612             7      ld  h,255 & (label1/256)
00000A 000A 2E34             7      ld  l,255 & label1
                                ;NOT は使わないので個人的にはどちらでも
00000C 000C **** E0004 **** 10      ld  hl,not label1
[EOF:test_high.asm:SHIFT_JIS]
```

**Comment by AILight** [comment, marked as answer]
high、low の演算子は見たことが無かったというのが本当のところです。式の中にhighとlowを入れて式の分解が上手く動かせるか？が焦点になるのですが、行けるか一度試してみたいと思います。

リンク切れはすみませんです。あとで直しておきたいと思います。

**Reply by SuperTurboZ** [reply]
high/lowのルーツは知りませんが古くはPROASM-IIはサポートしていました。pasmo や sjasm でも使えるようなのでそれほどマイナーではないのかなと思います。
AILZ80ASM専用に書く場合は優先順位などを気にしなくて済む.@h,.@lの方がすっきり書けてよいと思います。
よろしくお願いします。

**Reply by AILight** [reply]
high/low演算子の組み込みが完了しました。.@H, .@Lの実装があまり綺麗ではなかったのですが、演算子が作成されたおかげで、内部的にはhigh/low演算子を使って処理するように変更したところプログラムも非常に綺麗になりました。非常に助かったご指摘でした。ありがとうございました。

v0.9.9.23 でごご利用可能になりしたので、お使いいただけると嬉しく思います。

**Comment by tablacus** [comment]
`HIGH`演算子、`LOW`演算子の直後に括弧が存在する場合にエラーになってしまいますが、これを対策していただけるとありがたいです。
```asm
       0070                     CLR_PALET:
000070 0070 2182FF          10          ld      hl, TXTCUR + 2
000073 0073 **** E0004 ****  7          ld      b, HIGH(IOPALET + 100h)
```
これは https://github.com/meister68k/X1_compatible_rom/blob/master/X1_compatible_rom.z80 のソースをアセンブルした際に気付きました。

**Reply by AILight** [reply]
以下のissuesを作成して調査しました。こちらの不具合は既に対応済みでした。
https://github.com/AILight/AILZ80ASM/issues/370

**Comment by AILight** [comment]
皆様のおかげでよい機能を追加することができました。ありがとうございました。
